File: app/src/main/java/ml/docilealligator/infinityforreddit/readpost/ReadPostModification.java
package ml.docilealligator.infinityforreddit.readpost;

import java.util.concurrent.Executor;

import ml.docilealligator.infinityforreddit.RedditDataRoomDatabase;
import ml.docilealligator.infinityforreddit.account.Account;

public class ReadPostModification {
    public static void insertReadPost(RedditDataRoomDatabase redditDataRoomDatabase, Executor executor,
                                      String username, String postId, @ReadPostType int readPostType,
                                      int readPostsLimit) {
        executor.execute(() -> {
            ReadPostDao readPostDao = redditDataRoomDatabase.readPostDao();
            int limit = Math.max(readPostsLimit, 100);
            boolean isReadPostLimit = readPostsLimit != -1;
            while (readPostDao.getReadPostsCount(username, readPostType) > limit && isReadPostLimit) {
                readPostDao.deleteOldestReadPosts(username, readPostType);
            }
            if (username != null && !username.isEmpty()) {
                if (!redditDataRoomDatabase.accountDao().isAnonymousAccountInserted()) {
                    redditDataRoomDatabase.accountDao().insert(Account.getAnonymousAccount());
                }

                readPostDao.insert(new ReadPost(username, postId, readPostType));
            }
        });
    }

    public static void deleteReadPost(RedditDataRoomDatabase redditDataRoomDatabase, Executor executor,
                                      String username, String postId, @ReadPostType int readPostType) {
        executor.execute(() -> {
            if (username != null && !username.isEmpty()) {
                redditDataRoomDatabase.readPostDao().deleteReadPost(username, postId, readPostType);
            }
        });
    }
}

File: app/src/main/java/ml/docilealligator/infinityforreddit/readpost/ReadPostModification.kt
package ml.docilealligator.infinityforreddit.readpost

import ml.docilealligator.infinityforreddit.RedditDataRoomDatabase
import ml.docilealligator.infinityforreddit.account.Account
import kotlin.math.max

suspend fun insertReadPost(
    redditDataRoomDatabase: RedditDataRoomDatabase,
    username: String?,
    postId: String,
    @ReadPostType readPostType: Int,
    readPostsLimit: Int
) {
    if (!username.isNullOrEmpty()) {
        val readPostDaoKt = redditDataRoomDatabase.readPostDaoKt()
        val limit = max(readPostsLimit, 100)
        val isReadPostLimit = readPostsLimit != -1
        while (readPostDaoKt.getReadPostsCount(username, readPostType) > limit && isReadPostLimit) {
            readPostDaoKt.deleteOldestReadPosts(username, readPostType)
        }

        if (!redditDataRoomDatabase.accountDaoKt().isAnonymousAccountInserted()) {
            redditDataRoomDatabase.accountDaoKt().insert(Account.getAnonymousAccount())
        }

        readPostDaoKt.insert(ReadPost(username, postId, readPostType))
        if (readPostType == ReadPostType.ANONYMOUS_UPVOTED_POSTS) {
            readPostDaoKt.deleteReadPost(username, postId, ReadPostType.ANONYMOUS_DOWNVOTED_POSTS)
        } else if (readPostType == ReadPostType.ANONYMOUS_DOWNVOTED_POSTS) {
            readPostDaoKt.deleteReadPost(username, postId, ReadPostType.ANONYMOUS_UPVOTED_POSTS)
        }
    }
}

suspend fun deleteReadPost(
    redditDataRoomDatabase: RedditDataRoomDatabase,
    username: String?,
    postId: String,
    @ReadPostType readPostType: Int
) {
    if (!username.isNullOrEmpty()) {
        redditDataRoomDatabase.readPostDaoKt().deleteReadPost(username, postId, readPostType)
    }
}